kvmagent/kvmagent/plugins/zrm_plugin.py

```
@@ -0,0 +1,1234 @@
+                except Exception as e:
+                    err_str = str(e)
+                    # Retry with next candidate when device is not found or root node is required.
+                    if "Cannot find device" in err_str or "Need a root block node" in err_str:
+                        logger.debug("ZRM drive-mirror device=%s failed (%s), try next candidate" % (mirror_device, err_str[:80]))
+                        last_err = e
+                        continue
+                    last_err = e
+                    break
+            # If all drive-mirror candidates fail, diagnose block topology, then try the suggested root or blockdev fallback.
+            if last_err is not None:
+                suggested_root, topo_summary = self._diagnose_block_topology(vm_uuid, vol_uuid, primary_node or node_name)
```

Ruff (0.15.12)

[error] Undefined name `primary_node`

(F821)